host_modules/image_service.py

```diff
@@ -0,0 +1,135 @@
+"""
+This module provides services related to SONiC images, including:
+1) Downloading images
+2) Installing images
+3) Calculating checksums for images
+"""
+
+import errno
+import hashlib
+import logging
+import os
+import requests
+import stat
+import subprocess
+
+from host_modules import host_service
+import tempfile
+
+MOD_NAME = "image_service"
+
+DEFAULT_IMAGE_SAVE_AS = "/tmp/downloaded-sonic.bin"
+
+logger = logging.getLogger(__name__)
+
+
+class ImageService(host_service.HostModule):
+    """DBus endpoint that handles downloading and installing SONiC images"""
+
+    @host_service.method(
+        host_service.bus_name(MOD_NAME), in_signature="ss", out_signature="is"
+    )
+    def download(self, image_url, save_as):
+        """
+        Download a SONiC image.
+
+        Args:
+             image_url: url for remote image.
+             save_as: local path for the downloaded image. The directory must exist and be *all* writable.
+        """
+        logger.info("Download new sonic image from {} as {}".format(image_url, save_as))
+        # Check if the directory exists, is absolute and has write permission.
+        if not os.path.isabs(save_as):
+            logger.error("The path {} is not an absolute path".format(save_as))
+            return errno.EINVAL, "Path is not absolute"
+        dir = os.path.dirname(save_as)
+        if not os.path.isdir(dir):
+            logger.error("Directory {} does not exist".format(dir))
+            return errno.ENOENT, "Directory does not exist"
+        st_mode = os.stat(dir).st_mode
+        if (
+            not (st_mode & stat.S_IWUSR)
+            or not (st_mode & stat.S_IWGRP)
+            or not (st_mode & stat.S_IWOTH)
+        ):
+            logger.error("Directory {} is not all writable {}".format(dir, st_mode))
+            return errno.EACCES, "Directory is not all writable"
+        try:
+            response = requests.get(image_url, stream=True)
+            if response.status_code != 200:
+                logger.error(
+                    "Failed to download image: HTTP status code {}".format(
+                        response.status_code
+                    )
+                )
+                return errno.EIO, "HTTP error: {}".format(response.status_code)
+
+            with tempfile.NamedTemporaryFile(dir="/tmp", delete=False) as tmp_file:
+                for chunk in response.iter_content(chunk_size=8192):
+                    tmp_file.write(chunk)
+                temp_file_path = tmp_file.name
+            os.rename(temp_file_path, save_as)
```

Comment:
What will happen if this save_as already exists?

Reply:
os.rename will do nothing. Using os.replace instead.